blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc

```diff
@@ -0,0 +1,598 @@
+This might create some undesired coupling, though: the sending service must know which other services to invoke,
+and it also must be prepared for these services temporarily not being available.
+Service meshes such as https://istio.io/[Istio] come in helpful here, by providing capabilities like circuit breakers,
+request routing and much more.
+
+One general issue of such a synchronous approach is that it lacks re-playability,
+i.e. the possibility for new consumers to arrive after events have been sent and still be able to consume the entire event stream from the beginning.
+
+This can be achieved by using an asynchronous data exchange approach:
+by having the order service propagate events through a durable message log such as http://kafka.apache.org/[Apache Kafka],
+new consumers can be added as needed, enabling use cases you might not have had in mind originally,
+and without touching the source service.
+E.g. consider a data warehouse which should keep information about all the orders ever placed, or some full-text search functionality on orders based on https://www.elastic.co/products/elasticsearch[Elasticsearch].
+Once the purchase order events are in a Kafka topic
+(and depending on the topic's retention policy, they could remain their for an arbitrary long time or even indefinitely),
+new consumers can subscribe, process the topic from the very beginning and materialize a view of all the data in a microservice's database, search index, data warehouse etc.
+
+== The Issue of Dual Writes
+
+In order to provide their functionality, data-centered microservices will typically have their own local database.
+E.g. the order service may use a relational database to persist the information about purchase orders.
+When a new order is placed, this may result in an `INSERT` operation in a table `PurchaseOrder` in the service's own database.
+At the same time, the service may wish to send an event about the new order to Apache Kafka,
+so to propagate that information to other interested services.
+
+Simply issuing these two requests may lead to potential inconsistencies, though.
+The reason being that we cannot have one shared transaction that would span the service's database as well as Apache Kafka.
+So in unfortunate circumstances it might happen that we end up with having the new purchase order persisted in the local database,
+but not having sent the corresponding message to Kafka
+(e.g. due to some networking issue).
+Or, the other way around, we might have sent the message to Kafka but failed to persist the purchase order in the local database.
+Both situations are undesirable;
+this may cause no shipment to be created for a seemingly successfully placed order.
+Or a shipment gets created, but then there'd be no trace about the corresponding purchase order in the order service itself.
+
+So how can this situation be avoided?
```

Comment:
I think a short mention of XA and why it was not chosen should be added.

Reply:
Ok, I'll try. But "not chosen" sounds as if it were an option; but AFAICS it isn't, as Kafka cannot be enlisted in an XA transcation.

Reply:
Done.